pymc4/inference/sampling/sample.py

```diff
@@ -4,21 +4,21 @@
 
 
 def sample(model,
-           num_results=5000,
-           num_burnin_steps=3000,
+           num_results=1000,
+           num_burnin_steps=500,
            step_size=.4,
            num_leapfrog_steps=3,
            numpy=True):
     initial_state = []
-    for name, shape in model.unobserved.iteritems():
-        initial_state.append(.5 * tf.ones(shape, name="init_{}".format(name)))
+    for name in model.unobserved:
+        initial_state.append(.5 * tf.ones(model.unobserved[name].shape, name="init_{}".format(name)))
```

Comment:
This will create a lot of problems with namespace

Reply:
I do not know the way to go but to avot repeated calls of `self._f`

Reply:
``` python
unobserved = {}
        for i in self.variables:
            if self.variables[i] not in self.observed.values():
                unobserved[i] = self.variables[i]
        
        unobserved = collections.OrderedDict(unobserved)
        return unobserved
```

I could do this to avoid unobserved calling `f()` multiple times?

Reply:
``` python
In [2]: graph = tf.Graph()
   ...:
   ...: with graph.as_default():
   ...:     ed.Normal(0., 1.)
   ...:     print(graph.as_graph_def())
   ...:
node {
  name: "Normal/loc/input"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
        }
        float_val: 0.0
      }
    }
  }
}
node {
  name: "Normal/loc"
  op: "Identity"
  input: "Normal/loc/input"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal/scale/input"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
        }
        float_val: 1.0
      }
    }
  }
}
node {
  name: "Normal/scale"
  op: "Identity"
  input: "Normal/scale/input"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal_1/sample/sample_shape"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
          }
        }
      }
    }
  }
}
node {
  name: "Normal_1/sample/Normal/batch_shape_tensor/batch_shape"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
          }
        }
      }
    }
  }
}
node {
  name: "Normal_1/sample/concat/values_0"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 1
          }
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "Normal_1/sample/concat/axis"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "Normal_1/sample/concat"
  op: "ConcatV2"
  input: "Normal_1/sample/concat/values_0"
  input: "Normal_1/sample/Normal/batch_shape_tensor/batch_shape"
  input: "Normal_1/sample/concat/axis"
  attr {
    key: "N"
    value {
      i: 2
    }
  }
  attr {
    key: "T"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "Tidx"
    value {
      type: DT_INT32
    }
  }
}
node {
  name: "Normal_1/sample/random_normal/mean"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
        }
        float_val: 0.0
      }
    }
  }
}
node {
  name: "Normal_1/sample/random_normal/stddev"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
        }
        float_val: 1.0
      }
    }
  }
}
node {
  name: "Normal_1/sample/random_normal/RandomStandardNormal"
  op: "RandomStandardNormal"
  input: "Normal_1/sample/concat"
  attr {
    key: "T"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "seed"
    value {
      i: 0
    }
  }
  attr {
    key: "seed2"
    value {
      i: 0
    }
  }
}
node {
  name: "Normal_1/sample/random_normal/mul"
  op: "Mul"
  input: "Normal_1/sample/random_normal/RandomStandardNormal"
  input: "Normal_1/sample/random_normal/stddev"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal_1/sample/random_normal"
  op: "Add"
  input: "Normal_1/sample/random_normal/mul"
  input: "Normal_1/sample/random_normal/mean"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal_1/sample/mul"
  op: "Mul"
  input: "Normal_1/sample/random_normal"
  input: "Normal/scale"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal_1/sample/add"
  op: "Add"
  input: "Normal_1/sample/mul"
  input: "Normal/loc"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "Normal_1/sample/Shape"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 1
          }
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "Normal_1/sample/strided_slice/stack"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 1
          }
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "Normal_1/sample/strided_slice/stack_1"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 1
          }
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "Normal_1/sample/strided_slice/stack_2"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 1
          }
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "Normal_1/sample/strided_slice"
  op: "StridedSlice"
  input: "Normal_1/sample/Shape"
  input: "Normal_1/sample/strided_slice/stack"
  input: "Normal_1/sample/strided_slice/stack_1"
  input: "Normal_1/sample/strided_slice/stack_2"
  attr {
    key: "Index"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "T"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "begin_mask"
    value {
      i: 0
    }
  }
  attr {
    key: "ellipsis_mask"
    value {
      i: 0
    }
  }
  attr {
    key: "end_mask"
    value {
      i: 1
    }
  }
  attr {
    key: "new_axis_mask"
    value {
      i: 0
    }
  }
  attr {
    key: "shrink_axis_mask"
    value {
      i: 0
    }
  }
}
node {
  name: "Normal_1/sample/concat_1/axis"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "Normal_1/sample/concat_1"
  op: "ConcatV2"
  input: "Normal_1/sample/sample_shape"
  input: "Normal_1/sample/strided_slice"
  input: "Normal_1/sample/concat_1/axis"
  attr {
    key: "N"
    value {
      i: 2
    }
  }
  attr {
    key: "T"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "Tidx"
    value {
      type: DT_INT32
    }
  }
}
node {
  name: "Normal_1/sample/Reshape"
  op: "Reshape"
  input: "Normal_1/sample/add"
  input: "Normal_1/sample/concat_1"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "Tshape"
    value {
      type: DT_INT32
    }
  }
}
versions {
  producer: 26
}
```

For just one edward random variable there is some much change in the graph.

Reply:
That may be okay, if graph modifications are not frequently called (or it is hard to do) by user.